misc/pytorch_toolkit/chest_xray_screening_federated_gcn/src/utils/train_utils.py

```diff
@@ -0,0 +1,659 @@
+    backbone_wt=copy.deepcopy(backbone_model.state_dict())
+    fc_wt=copy.deepcopy(fc_layers.state_dict())
+    return prev_val, cnv_lyr1_wt,backbone_wt, fc_wt
+
+def initialize_training(site, img_pth, split_npz, train_transform, test_transform, b_sz, device):
+    
+    data_trn=construct_dataset(img_pth, split_npz, site, train_transform, tn_vl_idx=0)
+    trn_loader=DataLoader(data_trn,b_sz, shuffle=True, num_workers=1, pin_memory=False, drop_last=True)
+    
+    data_val=construct_dataset(img_pth, split_npz, site, test_transform, tn_vl_idx=1)
+    val_loader=DataLoader(data_val, 1, shuffle=False, num_workers=1, pin_memory=False, drop_last=True)
+    
+    criterion=Custom_Loss(site, device)
+    edge_index, edge_attr= compute_adjacency_matrix('confusion_matrix', site, split_npz)
+    
+    
+    return trn_loader, val_loader, criterion, edge_index, edge_attr
+
+
+
+def initialize_model_weights(cnv_lyr, backbone_model, fc_layers, gnn_model):
+    
+    cnv_wt=copy.deepcopy(cnv_lyr.state_dict())
+    backbone_wt=copy.deepcopy(backbone_model.state_dict())
+    fc_wt=copy.deepcopy(fc_layers.state_dict())
+    gnn_wt=copy.deepcopy(gnn_model.state_dict())
+    
+    return cnv_wt, backbone_wt, fc_wt, gnn_wt
+
+
+
+
+def save_model_weights(mx_nm, glbl_cnv_wt, glbl_backbone_wt, glbl_fc_wt, sit0_gnn_wt=None, sit1_gnn_wt=None, 
+                       sit2_gnn_wt=None,sit3_gnn_wt=None, sit4_gnn_wt=None):
+    torch.save({
+                'cnv_lyr_state_dict': glbl_cnv_wt,
+                'backbone_model_state_dict': glbl_backbone_wt,
+                'fc_layers_state_dict': glbl_fc_wt,
+                'sit0_gnn_model': sit0_gnn_wt,
+                'sit1_gnn_model': sit1_gnn_wt,
+                'sit2_gnn_model': sit2_gnn_wt,
+                'sit3_gnn_model': sit3_gnn_wt,
+                'sit4_gnn_model': sit4_gnn_wt,
+                }, mx_nm)
+    
+    
+def instantiate_architecture(ftr_dim, model_name, gnn=False):
+    # If gnn=True, then instantiate the GNN architecture
+  
+    if model_name=='densenet':
+        inp_dim=1024
+        backbone_model=models.densenet121(pretrained=True)
+        backbone_model.classifier=nn.Identity()
+    elif model_name=='resnet':
+        inp_dim=512
+        backbone_model=models.resnet18(pretrained=True)
+        backbone_model.fc=nn.Identity()
+    
+    elif model_name=='xception':
+        inp_dim=2048
+        backbone_model=xception.xception(pretrained=True)
+        backbone_model.fc=nn.Identity()
+    
+
+    cnv_lyr=First_Conv()
+    fc_layers=Fully_Connected_Layer(inp_dim, ftr_dim)
+    if gnn:
+        gnn_model=GNN_Network(in_chnls=512, base_chnls=1, grwth_rate=1, depth=1, aggr_md='mean', ftr_dim=4)
+        return cnv_lyr, backbone_model, fc_layers, gnn_model
+    
+    return cnv_lyr, backbone_model, fc_layers
+
+
+#Main function for training                
+def trainer_with_GNN(lr, b_sz, img_pth, split_npz, train_transform, test_transform,
```

Comment:
Very large file. Can this be broken into multiple files? Maybe one file which contains code for the trainer with gnn and another for without gnn.

Reply:
This needs to be resolved.